src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,231 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <uv.h>
+
+#include "nvim/api/private/helpers.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/lua/executor.h"
+#include "nvim/vim.h"
+#include "nvim/version.h"
+
+#ifdef __linux__
+
+#define XLIB_NAME "libX11.so.6"
+typedef void *_XDisplay;
+typedef struct {
+    char data[18];
+} _XkbStateRec;
+typedef int _XStatus;
+
+enum { kLockedModsOffset = 9 };
+enum { _XkbUseCoreKbd = 0x0100 };
+#endif
+
+#ifdef WIN32
+#include <Winuser.h>
+#endif
+
+#ifdef __APPLE__
+#define Boolean Boolean_I_Dont_Care
+#include <IOKit/IOKitLib.h>
+#include <IOKit/IOReturn.h>
+#include <IOKit/hidsystem/IOHIDLib.h>
+#include <IOKit/hidsystem/IOHIDParameter.h>
+#include <CoreFoundation/CoreFoundation.h>
+#undef Boolean
+#endif
+
+enum {
+    kNumLock = 0x10,
+    kCapsLock = 0x02,
+};
+
+
+#if defined(__linux__)
+static bool xlib_opened = false;
+static uv_lib_t xlib;
+static _XDisplay *(*_XOpenDisplay)(void *);
+static int (*_XCloseDisplay)(_XDisplay *);
+static _XStatus (*_XkbGetState)(_XDisplay *, unsigned, _XkbStateRec *);
+
+/// A helper function to be used to load the libX11 and required functions
+static bool os_xlib_init(const char **const err)
+{
+  if (xlib_opened) {
+    return true;
+  }
+  const int udl_ret = uv_dlopen(XLIB_NAME, &xlib);
+  if (udl_ret != 0) {
+    *err = "Failed to open library libX11";
+  } else {
+    xlib_opened = true;
+  }
+
+  if ((uv_dlsym(&xlib, "XOpenDisplay", (void **)&_XOpenDisplay)
+       || uv_dlsym(&xlib, "XCloseDisplay", (void **)&_XCloseDisplay)
+       || uv_dlsym(&xlib, "XkbGetState", (void **)&_XkbGetState)) != 0) {
+    *err = "Unable to get the functions of X11";
+    return false;
+  }
+  return xlib_opened;
+}
+
+/// Get information about Num/Caps Lock state in linux
```

Comment:
This function gets state for scroll lock as well.

Reply:
No, the scroll lock is not working in MacOS, so shouldn't we drop it, as you said?

Reply:
I only said to drop it *on Mac OS X*. Not *everywhere*.